Ensemble Selection Method Customization in TabRepo

alee4977 opened the pull request:
**Summary of Changes**

This pull request introduces enhancements to the TabRepo repository to allow users to select the ensemble selection method. 

**Detailed Changes**

1. **EvaluationRepository Class (EvaluationRepository)**:

    * Modified the `evaluate_ensemble()` function to take in an `ensemble_method_name` argument:
        * **Purpose:** Integrated support for selecting the ensemble selection method used.
        * **Functionality:** Passes the ensemble method name to the _construct_ensemble_selection_config_scorer method that constructs a scorer object.
    * Modified the `_construct_ensemble_selection_config_scorer()` function to take in an `ensemble_method_name` argument:
        * **Purpose:** Support for selecting the ensemble selection method used.
        * **Functionality:** Passes the ensemble method name during creation of the `EnsembleSelectionConfigScorer` object.

2. **EnsembleSelectionConfigScorer Class (EnsembleSelectionConfigScorer)**:

    * Added `ensemble_method_name` attribute to the `EnsembleSelectionConfigScorer` class: 
        * **Purpose:** The scorer object evaluates the configurations via simulating ensemble selection. This customization allows the user to select the ensemble method used to construct the ensemble. 
        * **Functionality:** Asserts that the `ensemble_method_name` passed in is one of the five options available. Sets the object’s attribute `ensemble_method_name` to the name passed. Updates the `ensemble_selection_kwargs` to include the ensemble method name.

3. **EnsembleScorer Class (EnsembleScorer)**:

    * Set the default `ensemble_method` to `EnsembleSelectionMethod` class:
        * **Purpose:** This simplifies the code by having the logic for ensemble selection in one class, instead of multiple classes. 
    * Added an attribute `ensemble_method_name` to the `EnsembleScorer` class so that the fit method is based on the value of the `ensemble_method_name`.
        * **Purpose:** The `EnsembleSelectionMethod` has access to the ensemble method name passed via keyword arguments.

**Impact**

* **Customizability:** Allows the user to specify the ensemble selection method.
* **Simplicity:** All the logic for the different ensemble methods are in one class. 

**Results**

When running the `evaluate_ensemble()` method on the given datasets and configurations, the weight vector for the ensemble model that is returned found using the ensemble selection method specified by the user.

prateekdesai04 commented on tabrepo/repository/evaluation_repository.py:
```diff
@@ -269,6 +269,7 @@ def labels_val(self, dataset: str, fold: int) -> np.array:
 
     def dataset_metadata(self, dataset: str) -> dict:
         metadata = self._df_metadata[self._df_metadata["dataset"] == dataset]
+        print(metadata)
```
Why do we need this ?

prateekdesai04 commented on tabrepo/repository/evaluation_repository.py:
```diff
@@ -315,6 +316,7 @@ def evaluate_ensemble(
         datasets: List[str],
         configs: List[str] = None,
         *,
+        ensemble_method_name: str = "greedy ensemble selection",
```
Does it seem like a good idea to have strings defined like this ?
do not use spaces

prateekdesai04 commented on tabrepo/repository/evaluation_repository.py:
```diff
@@ -379,13 +383,15 @@ def _construct_config_scorer(self,
 
     def _construct_ensemble_selection_config_scorer(self,
                                                     ensemble_size: int = 10,
+                                                    ensemble_method_name: str = "greedy ensemble selection",
```
ditto

prateekdesai04 commented on tabrepo/simulation/ensemble_selection_config_scorer.py:
```diff
@@ -21,7 +21,7 @@
 
 @ray.remote
-def compute_error_ray(config_scorer, configs: List[str], task: str) -> (float, dict):
+def compute_error_ray(config_scorer, configs: List[str], task: str) -> (float, dict): # type: ignore
```
is this suppression due to the debugger in IDE?
It is not needed here

prateekdesai04 commented on tabrepo/simulation/ensemble_selection_config_scorer.py:
```diff
@@ -131,10 +132,16 @@ def evaluate_task(self, dataset: str, fold: int, models: List[str]) -> Tuple[flo
+
+        if eval_metric.needs_pred:
+            y_val_pred = weighted_ensemble.predict(pred_val)
+        else:
```
could you give more context here

prateekdesai04 commented on tabrepo/simulation/ensemble_selection_config_scorer.py:
```diff
@@ -148,8 +155,9 @@ def __init__(self,
+                 ensemble_method_name: str = "greedy ensemble selection",
                  ensemble_selection_kwargs=None,
-                 backend: str = 'native',
+                 backend: str = 'ray',
```
why do we switch to ray ?
more context here would be good

prateekdesai04 commented on tabrepo/simulation/ensemble_selection_config_scorer.py:
```diff
@@ -185,6 +194,8 @@ def __init__(self,
         self.ranker = ranker
         self.tid_to_dataset_name_dict = tid_to_dataset_name_dict
         self.ensemble_size = ensemble_size
+        assert ensemble_method_name in ["greedy ensemble selection", "cmaes", "cmaes with normalization", "quality optimization", "quality diversity optimization"]
```
ditto, refer to first comment

alee4977 commented:
**Summary of Changes**
This pull request introduces changes to the `baselines.py` file in TabRepo to allow the specification of the ensemble selection method used to run experiments. Key modifications include changes to multiple functions in the `baselines.py` file. 

**Detailed Changes**

**1. baselines.py file:**

* Modified the `evaluate_configs()` function:
    * Purpose: This method evaluates all the ensembles created on the different task and fold combinations, and returns the errors, normalized errors, times to train, and inference times as a row. Now the method used to create the ensemble can be specified. 
    * Functionality: The `ensemble_method_name` passed in is then passed as a parameter to the `evaluate_ensemble()` method called within the function. 
* Modified the `framework_default_results()` function:
    * Purpose: The function calls `evaluate_configs()`, which takes in `ensemble_method_name` as an parameter. 
* Modified the `evaluate_tid()` function:
    * Purpose: The function calls `evaluate_configs()`, which takes in `ensemble_method_name` as a parameter. 
* Modified the `zeroshot_name()` function:
    * Purpose: This function returns the name of the zeroshot method based on different parameters, and given the ensemble selection method is also a parameter, this method was updated. 
    * Functionality: The name of the ensemble selection method is added at the end of the zeroshot method name. 
* Modified the `zeroshot_results()` function:
    * Purpose: This method runs a zeroshot experiment, but now with the option of specifying the ensemble selection method. 
    * Functionality: This function takes in the `ensemble_method_name` parameter and passes it to the `evaluate_configs()` method. 

**Impact**

* Customizability: Allows users to customize the ensemble selection method used to build the ensemble model. 

**Results**
After the introduction of these changes, users of TabRepo will be able to specific the ensemble selection method used to build ensembles when running experiments. This is a useful feature that allows users to compare the performance of ensembles built using different methods.
